### DISCUSSION TITLE
global return and named module

### DISCUSSION THREAD

**Discussion by codehz** [edubart/nelua-lang#102] [Ideas]
In lua, we can return value from a lua file, but it is impossible in nelua for now (because require return void, and cannot return type so it cannot be used as module namespace).

# Idea: make this valid

mymodule.nelua:
```lua
local module = @record{}

function module.method() end

-- make global return can return anything (comptime or not)
return module
```

main.nelua:
```lua
local mod = require 'mymodule'

mod.method()
```

## Alternative syntax(syntax sugar)

mymodule.nelua:
```lua
module function method() end
```

**Comment by edubart** [comment]
This is on the roadmap, and definitely will be done, it will be just like as Lua (as you first example). I expect this to be one of the next new language features.

**Reply by codehz** [reply]
the first syntax has a problem, you can only attach function to @record, since `global module.INFO: string` syntax is only available in global variable...

**Reply by edubart** [reply]
`global module.INFO: string` there is fine even if `module` is local, the `global` in that context is a "global" variable on the record type, not a global in the sense of the global scope, this is because the grammar needs a keyword before a declaration so the compiler knows that a variable is being declared, this solves many issues otherwise the grammar would have ambiguities with assignment, that is, while in Lua you can declare a variable without a keyword, in Nelua you can't, you always need `global` or `local` preceding a declaration, and I've choose `global` instead of `local` to explicitly declare variables inside a record because for me made more sense than `local`.

So the module system would be like:

```lua
local module = @record{}

-- A module function.
function module.f() end

-- A module variable.
global module.INFO: string
```

Then using:
```lua
local mod = require 'module'

-- Call module function
mod.f()

-- Use module variable
print(mod.INFO)
```

**Reply by codehz** [reply]
Ah, this explanation is very clear, thanks

**Reply by codehz** [reply]
BTW, why function module.f doesn't need any prefix? there are some inconsistencies in the grammar here

**Comment by edubart** [comment]
Just updating, modules system has been implemented recently, you can now return types from modules, it works as the first idea.
